lib/nghttp2_buf.c

```diff
@@ -223,13 +223,60 @@ int nghttp2_bufs_wrap_init(nghttp2_bufs *bufs, uint8_t *begin, size_t len,
   return 0;
 }
 
+int nghttp2_bufs_wrap_init2(nghttp2_bufs *bufs, uint8_t **bufs_in, size_t in_len,
+                          size_t buf_len, nghttp2_mem *mem) {
+  size_t i = 0;
+  nghttp2_buf_chain *cur_chain;
+  nghttp2_buf_chain *pre_chain = NULL;
+  nghttp2_buf_chain *head_chain = NULL;
+
+  for (i = 0; i < in_len; ++i) {
+    uint8_t *begin = bufs_in[i];
+
+    cur_chain = nghttp2_mem_malloc(mem, sizeof(nghttp2_buf_chain));
+    if (cur_chain == NULL) {
+      while (head_chain) {
+        nghttp2_mem_free(mem, head_chain);
+        head_chain = head_chain->next;
```

Comment:
head_chain was already freed.  Accessing head_chain->next is heap-use-after-free.